sql/transactions/transactions.go

```diff
@@ -0,0 +1,185 @@
+func Applied(db sql.Executor, id types.TransactionID) error {
+	if rows, err := db.Exec("update transactions set applied = true where id = ?1 returning id", func(stmt *sql.Statement) {
+		stmt.BindBytes(1, id.Bytes())
+	}, nil); err != nil {
+		return fmt.Errorf("applied %s: %w", id, err)
+	} else if rows == 0 {
+		return fmt.Errorf("%w: tx %s", sql.ErrNotFound, id)
+	}
+	return nil
+}
+
+// Delete transaction from database.
+func Delete(db sql.Executor, id types.TransactionID) error {
+	if _, err := db.Exec("delete from transactions where id = ?1",
+		func(stmt *sql.Statement) {
+			stmt.BindBytes(1, id.Bytes())
+		}, nil); err != nil {
+		return fmt.Errorf("delete %s: %w", id, err)
+	}
+	return nil
+}
+
+// tx, layer, block, origin.
+func decodeTransaction(id types.TransactionID, stmt *sql.Statement) (*types.MeshTransaction, error) {
+	var (
+		tx     types.Transaction
+		origin types.Address
+		bid    types.BlockID
+	)
+	if _, err := codec.DecodeFrom(stmt.ColumnReader(0), &tx); err != nil {
+		return nil, fmt.Errorf("decode %w", err)
+	}
+	lid := types.NewLayerID(uint32(stmt.ColumnInt64(1)))
+	stmt.ColumnBytes(2, bid[:])
+	stmt.ColumnBytes(3, origin[:])
+	tx.SetOrigin(origin)
+	tx.SetID(id)
+	return &types.MeshTransaction{
+		Transaction: tx,
+		LayerID:     lid,
+		BlockID:     bid,
+	}, nil
+}
+
+// Get transaction from database.
+func Get(db sql.Executor, id types.TransactionID) (tx *types.MeshTransaction, err error) {
+	if rows, err := db.Exec("select tx, layer, block, origin from transactions where id = ?1",
+		func(stmt *sql.Statement) {
+			stmt.BindBytes(1, id.Bytes())
+		}, func(stmt *sql.Statement) bool {
+			tx, err = decodeTransaction(id, stmt)
+			return true
+		}); err != nil {
+		return nil, fmt.Errorf("get %s: %w", id, err)
+	} else if rows == 0 {
+		return nil, fmt.Errorf("%w: tx %s", sql.ErrNotFound, id)
+	}
+	return tx, err
+}
+
+// GetBlob loads transaction as an encoded blob, ready to be sent over the wire.
+func GetBlob(db sql.Executor, id types.TransactionID) (buf []byte, err error) {
+	if rows, err := db.Exec("select tx from transactions where id = ?1",
+		func(stmt *sql.Statement) {
+			stmt.BindBytes(1, id.Bytes())
+		}, func(stmt *sql.Statement) bool {
+			buf = make([]byte, stmt.ColumnLen(0))
+			stmt.ColumnBytes(0, buf)
+			return true
+		}); err != nil {
+		return nil, fmt.Errorf("get %s: %w", id, err)
+	} else if rows == 0 {
+		return nil, fmt.Errorf("%w: tx %s", sql.ErrNotFound, id)
+	}
+	return buf, nil
+}
+
+// Has returns true if transaction is stored in the database.
+func Has(db sql.Executor, id types.TransactionID) (bool, error) {
+	rows, err := db.Exec("select 1 from transactions where id = ?1",
+		func(stmt *sql.Statement) {
+			stmt.BindBytes(1, id.Bytes())
+		}, nil)
+	if err != nil {
+		return false, fmt.Errorf("has %s: %w", id, err)
+	}
+	return rows > 0, nil
+}
+
+// query MUST ensure that this order of fields tx, layer, block, origin, id.
```

Comment:
prolly safer to define the select stmt as a string. `select tx, layer, block, origin, id`